autoretrieve/engine.go

```diff
@@ -0,0 +1,524 @@
+			prevMetadata, err := e.getKeyMetadataMap(ctx, contextID)
+			if err != nil {
+				if err != datastore.ErrNotFound {
+					return cid.Undef, fmt.Errorf("could not get metadata for context id: %s", err)
+				}
+				log.Warn("No metadata for existing context ID, generating new advertisement")
+			}
+
+			if md.Equal(prevMetadata) {
+				// Metadata is the same; no change, no need for new advertisement.
+				return cid.Undef, provider.ErrAlreadyAdvertised
+			}
+
+			// Linked list is the same, but metadata is different, so generate
+			// new advertisement with same linked list, but new metadata.
+			cidsLnk = cidlink.Link{Cid: c}
+		}
+
+		if err = e.putKeyMetadataMap(ctx, contextID, &md); err != nil {
+			return cid.Undef, fmt.Errorf("failed to write context id to metadata mapping: %s", err)
+		}
+	} else {
+		log.Info("Creating removal advertisement")
+
+		if c == cid.Undef {
+			return cid.Undef, provider.ErrContextIDNotFound
+		}
+
+		// And if we are removing it means we probably do not have the list of
+		// CIDs anymore, so we can remove the entry from the datastore.
+		err = e.deleteKeyCidMap(ctx, contextID)
+		if err != nil {
+			return cid.Undef, fmt.Errorf("failed to delete context id to entries cid mapping: %s", err)
+		}
+		err = e.deleteCidKeyMap(ctx, c)
+		if err != nil {
+			return cid.Undef, fmt.Errorf("failed to delete entries cid to context id mapping: %s", err)
+		}
+		err = e.deleteKeyMetadataMap(ctx, contextID)
+		if err != nil {
+			return cid.Undef, fmt.Errorf("failed to delete context id to metadata mapping: %s", err)
+		}
+
+		// Create an advertisement to delete content by contextID by specifying
+		// that advertisement has no entries.
+		cidsLnk = schema.NoEntries
+
+		// The advertisement still requires a valid metadata even though
+		// metadata is not used for removal.  Create a valid empty metadata.
+		md = metadata.New(metadata.Bitswap{})
+	}
+
+	mdBytes, err := md.MarshalBinary()
+	if err != nil {
+		return cid.Undef, err
+	}
+
+	adv := schema.Advertisement{
+		Provider:  providerID,
+		Addresses: addresses,
+		Entries:   cidsLnk,
+		ContextID: contextID,
+		Metadata:  mdBytes,
+		IsRm:      isRm,
+	}
+
+	// Get the previous advertisement that was generated
+	prevAdvID, err := e.getLatestAdCid(ctx)
+	if err != nil {
+		return cid.Undef, fmt.Errorf("could not get latest advertisement: %s", err)
+	}
+
+	// Check for cid.Undef for the previous link. If this is the case, then
+	// this means there is a "cid too short" error in IPLD links serialization.
+	if prevAdvID != cid.Undef {
+		log.Info("Latest advertisement CID was undefined - no previous advertisement")
+		prev := ipld.Link(cidlink.Link{Cid: prevAdvID})
+		adv.PreviousID = &prev
+	}
+
+	// Sign the advertisement.
+	if err := adv.Sign(e.key); err != nil {
+		return cid.Undef, err
+	}
+	return e.Publish(ctx, adv)
+}
+
+func (e *AutoretrieveEngine) putKeyCidMap(ctx context.Context, contextID []byte, c cid.Cid) error {
```

Comment:
maybe move all the store CRUD methods to a `datastore.go` file?

Reply:
Same as above, https://github.com/filecoin-project/index-provider/